handlers/definitionsForRoles.js

```diff
@@ -27,86 +27,87 @@ module.exports = {
-            // all the ABDefinition.id that need to be exported.
-
-            var hashKey = roleIDs.join(",");
-            if (!hashIDs[hashKey] || hashIDs[hashKey].length == 0) {
-               req.log("building ID hash");
-               req.performance.mark("buildIDHash");
-               var applications = AB.applications(
-                  (a) => isSystemUser || a.isAccessibleForRoles(roles)
-               );
-
-               req.log(
-                  `definition_manager.definitionsForRoles: found ${applications.length} applications to export`
-               );
-
-               // This takes a long time!
-               // Cache this?
-               let aIDs = [];
-               applications.forEach((a) => {
-                  a.exportIDs(aIDs);
-               });
-
-               // NOTE: we also need to make sure all the System Objects in the definitions.
-               let systemObjects = AB.objects((o) => o.isSystemObject);
-               for (var i = 0; i < systemObjects.length; i++) {
-                  systemObjects[i].exportIDs(aIDs);
-               }
-
-               hashIDs[hashKey] = aIDs;
-
-               AB.cache("defs-for-role", hashIDs);
-               req.performance.measure("buildIDHash");
-            }
-
-            ids = hashIDs[hashKey];
-            req.log(
-               `definition_manager.definitionsForRoles: found ${ids.length} ids to export.`
-            );
-            var definitions = [];
-            ids.forEach((id) => {
-               let def = AB.definitionByID(id, true);
-               if (def) {
-                  definitions.push(def);
-               }
-            });
-
-            cb(null, definitions);
-         })
-         .catch((err) => {
-            // we clear the cache just in case our data was incorrect.
-            // AB.cacheClear(ServiceKey);
-            req.notify.developer(err, {
-               context:
-                  "Service:definition_manager.definitionsForRoles: Error initializing ABFactory",
-            });
-            cb(err);
+      try {
+         const AB = await ABBootstrap.init(req);
+         const stringifiedDefs = AB.cache("cached-defs") || {};
+         const hashKey = roleIDs.join(",");
+
+         if (stringifiedDefs[hashKey] != null) {
+            cb(null, stringifiedDefs[hashKey]);
+
+            return;
+         }
+
+         req.log("building ID hash");
+         req.performance.mark("buildIDHash");
+
+         const isSystemUser =
+            roleIDs.filter((roleId) => AB.defaultSystemRoles().includes(roleId))
+               .length > 0;
+         const applications = AB.applications(
+            (a) => isSystemUser || a.isAccessibleForRoles(roles)
+         );
+
+         req.log(
+            `definition_manager.definitionsForRoles: found ${applications.length} applications to export`
+         );
+
+         // This takes a long time!
+         // Cache this?
+         stringifiedDefs[hashKey] = [];
```

Comment:
ok, looks like we are being tricky and using `stringifiedDefs[hashKey]` to mean different things in the process of this function.  While it looks like we are correctly managing it, I have a feeling that this can be confusing to someone else coming in and trying to read the code.

Let's keep `stringifiedDefs[hashKey]` to only hold the final cached definitions.

In these intermediate steps of gathering the ids, use another variable: `allIDs` or something.

so
```
const allIDs = [];

applications.forEach((a) => { a.exportIDs(allIDs); } );

...

stringifiedDefs[hashKey] = await req.worker( (defs) => JSON.stringify(defs), [ 
   allIDs.map().filter()
]);
```

Reply:
Thank you! Considering the worker's blocking during `JSON` data transfer, should I use `bfj.stringify` instead of the `worker`? Also, do you think I should retain the method in our ab-utils?

Reply:
https://github.com/digi-serve/ab_service_definition_manager/pull/64/commits/f0f74aa310a76735ef3d444bee2a1aa0c3bcfc57

Reply:
in the case of the worker blocking, in this case it should be ok.  Since we are spawning a new thread for each call to `req.worker()` it won't affect the performance of the other threads.  So in that thread, we can use the faster `JSON.stringify()` instead of the `bfj.stringify()` command.

Let's keep the `req.worker()` command in `ab-utils`.  That can be a nice feature now for all the other services that try to perform CPU intensive tasks.  For example, any of our image resizing operations should now take advantage of `req.worker()` if we can.  

Your latest changes look great!  Let's commit this!